pkg/envoy/cds/builder.go

```diff
@@ -28,9 +31,117 @@ import (
 	"github.com/openservicemesh/osm/pkg/trafficpolicy"
 )
 
+type clusterBuilder struct {
+	proxyIdentity identity.ServiceIdentity
+
+	outboundMeshTrafficClusterConfigs []*trafficpolicy.MeshClusterConfig
+
+	inboundMeshTrafficClusterConfigs []*trafficpolicy.MeshClusterConfig
+
+	egressTrafficClusterConfigs []*trafficpolicy.EgressClusterConfig
+
+	sidecarSpec configv1alpha2.SidecarSpec
+
+	egressEnabled bool
+
+	metricsEnabled bool
+
+	envoyTracingAddress *xds_core.Address
+}
+
+func NewClusterBuilder() *clusterBuilder { //nolint: revive // unexported-return
+	return &clusterBuilder{}
+}
+
+func (b *clusterBuilder) SetProxyIdentity(identity identity.ServiceIdentity) *clusterBuilder {
+	b.proxyIdentity = identity
+	return b
+}
+
+func (b *clusterBuilder) SetSidecarSpec(spec configv1alpha2.SidecarSpec) *clusterBuilder {
+	b.sidecarSpec = spec
+	return b
+}
+
+func (b *clusterBuilder) SetOutboundMeshTrafficClusterConfigs(outboundMeshTrafficClusterConfigs []*trafficpolicy.MeshClusterConfig) *clusterBuilder {
+	b.outboundMeshTrafficClusterConfigs = outboundMeshTrafficClusterConfigs
+	return b
+}
+
+func (b *clusterBuilder) SetInboundMeshTrafficClusterConfigs(inboundMeshTrafficClusterConfigs []*trafficpolicy.MeshClusterConfig) *clusterBuilder {
+	b.inboundMeshTrafficClusterConfigs = inboundMeshTrafficClusterConfigs
+	return b
+}
+
+func (b *clusterBuilder) SetEgressTrafficClusterConfigs(egressTrafficClusterConfigs []*trafficpolicy.EgressClusterConfig) *clusterBuilder {
+	b.egressTrafficClusterConfigs = egressTrafficClusterConfigs
+	return b
+}
+
+func (b *clusterBuilder) SetEgressEnabled(egressEnabled bool) *clusterBuilder {
+	b.egressEnabled = egressEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetMetricsEnabled(metricsEnabled bool) *clusterBuilder {
+	b.metricsEnabled = metricsEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetEnvoyTracingAddress(tracingAddress *xds_core.Address) *clusterBuilder {
+	b.envoyTracingAddress = tracingAddress
+	return b
+}
+
+func (b *clusterBuilder) tracingEnabled() bool {
+	return b.envoyTracingAddress != nil
+}
+
 // replacer used to configure an Envoy cluster's altStatName
 var replacer = strings.NewReplacer(".", "_", ":", "_")
 
+func (b *clusterBuilder) Build() ([]types.Resource, error) {
+	var clusters []*xds_cluster.Cluster
+
+	// Build upstream clusters based on allowed outbound traffic policies
+	if b.outboundMeshTrafficClusterConfigs != nil {
+		clusters = append(clusters, b.buildUpstreamClusters()...)
+	}
+
+	// Build local clusters based on allowed inbound traffic policies
+	if b.inboundMeshTrafficClusterConfigs != nil {
+		clusters = append(clusters, b.buildLocalClusters()...)
+	}
+
+	// Add egress clusters based on applied policies
+	if b.egressTrafficClusterConfigs != nil {
+		clusters = append(clusters, b.getEgressClusters()...)
+	}
+
+	// Add an outbound passthrough cluster for egress if global mesh-wide Egress is enabled
```

Comment:
note the change in logic here: https://github.com/openservicemesh/osm/blob/main/pkg/envoy/cds/response.go#L47-L57

Prev:
- get outbound passthru cluster + handle error
- check if egress enabled - if yes, add outboundpassthroughcluster

Now:
- check if egress enabled - if yes, only then try to get outbound passthru cluster and handle error